File: packages/flutter_amazon_connect_webrtc/ios/Classes/ConnectVideoView.swift
import AmazonChimeSDK
import Flutter
import UIKit

/// Factory for the video-tile PlatformView. Hosts Chime's `DefaultVideoRenderView` and binds it to
/// the requested tile via the ``ChimeCallManager``.
class ConnectVideoViewFactory: NSObject, FlutterPlatformViewFactory {
    private let callManager: ChimeCallManager

    init(callManager: ChimeCallManager) {
        self.callManager = callManager
        super.init()
    }

    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        ConnectVideoPlatformView(frame: frame, args: args as? [String: Any], callManager: callManager)
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        FlutterStandardMessageCodec.sharedInstance()
    }
}

class ConnectVideoPlatformView: NSObject, FlutterPlatformView {
    private let renderView: DefaultVideoRenderView
    private let tileId: Int
    private weak var callManager: ChimeCallManager?

    init(frame: CGRect, args: [String: Any]?, callManager: ChimeCallManager) {
        self.tileId = (args?["tileId"] as? Int) ?? -1
        self.callManager = callManager
        self.renderView = DefaultVideoRenderView(frame: frame)
        super.init()

        if let mirror = args?["mirror"] as? Bool {
            renderView.mirror = mirror
        }
        if tileId >= 0 {
            callManager.bindVideoView(renderView, tileId: tileId)
        }
    }

    func view() -> UIView { renderView }

    deinit {
        if tileId >= 0 {
            callManager?.unbindVideoView(tileId: tileId)
        }
    }
}

File: packages/flutter_amazon_connect_webrtc/android/src/main/kotlin/com/chimeflutter/connect_webrtc/ConnectVideoView.kt
package com.chimeflutter.connect_webrtc

import android.content.Context
import android.view.View
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.DefaultVideoRenderView
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/** Factory for the video-tile PlatformView; hosts Chime's [DefaultVideoRenderView]. */
class ConnectVideoViewFactory(private val callManager: ChimeCallManager) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val params = args as? Map<String, Any?> ?: emptyMap()
        return ConnectVideoPlatformView(context, params, callManager)
    }
}

class ConnectVideoPlatformView(
    context: Context,
    params: Map<String, Any?>,
    private val callManager: ChimeCallManager,
) : PlatformView {

    private val tileId = (params["tileId"] as? Int) ?: -1
    private val renderView = DefaultVideoRenderView(context)

    init {
        renderView.mirror = (params["mirror"] as? Boolean) ?: false
        if (tileId >= 0) callManager.bindVideoView(renderView, tileId)
    }

    override fun getView(): View = renderView

    override fun dispose() {
        if (tileId >= 0) callManager.unbindVideoView(tileId)
    }
}
